File: C/other/ternarySearch.c
#include <stdio.h> 
int ternarySearch(int l, int r, int key, int ar[]) 
{ 
    if (r >= l) { 
        int mid1 = l + (r - l) / 3; 
        int mid2 = r - (r - l) / 3; 
        if (ar[mid1] == key) { 
            return mid1; 
        } 
        if (ar[mid2] == key) { 
            return mid2; 
        }
        if (key < ar[mid1]) { 
            return ternarySearch(l, mid1 - 1, key, ar); 
        } 
        else if (key > ar[mid2]) { 
            return ternarySearch(mid2 + 1, r, key, ar); 
        } 
        else { 
            return ternarySearch(mid1 + 1, mid2 - 1, key, ar); 
        } 
    } 
    return -1; 
} 

int main() 
{ 
    int num;
    printf("Enter number of elements: ");
    scanf("%d",&num);
    printf("\nEnter %d sorted elements: ",num);
    int array[num];
    for(int i = 0;i<num;i++){
        scanf("%d",&array[i]);
    }
    int key, index;
    printf("\nEnter Key to be searched: ");
    scanf("%d",&key);
    index = ternarySearch(0, num, key, array); 
    printf("Index of %d is %d\n", key, index); 
} 
/*
Time Complexity: O(log3 n)
Space Complexity: O(1)

Enter number of elements: 8

Enter 8 elements: 12 25 48 52 67 78 89 91

Enter Key to be searched: 52
Index of 52 is 3
*/

File: Julia/search/ternary_search.jl
"""Julia program to implement Ternary Search algorithm.
Ternary Search is a Divide and Conquer strategy based algorithm similar to binary search.
Here we divide the entire array into three parts and iteratively search for the required element
"""

function ternary_search(arr, low, high, ele)
    while (high >= low)
        # FInd the two required midpoints to divide the array into three.
        midA = low + (high - low) ÷ 3
        midB = high - (high - low) ÷ 3
        # Search, if the element is present in the midpoint indices
        if (arr[midA] == ele)
            return true
        elseif (arr[midB] == ele)
            return true
        end
    
        # Identify the part in which the required element belongs to 
        if (ele < arr[midA])
            high = midA - 1
        elseif (ele > arr[midB])
            low = midB + 1
        else
            low = midA + 1
            high = midB - 1
        end
    end
    return false
end


print("How many numbers are present in the array? ")
n = readline()
n = parse(Int, n)
if (n <= 0)
    println("Array is Empty!!!")
    exit()
end
arr = Int[]
print("Enter the numbers: ")
arr = [parse(Int, num) for num in split(readline())] 
print("Which number do you want to search in the array? ")
ele = readline()
ele = parse(Int, ele)
# Sort the array in ascending order
arr = sort(arr)
res = ternary_search(arr, 1, n, ele)
if (res == 0)
    print("The number $ele is not present in the array")
else
    print("The number $ele is present in the array.")
end


"""
Time Complexity - O(log(n)), where 'n' is the size of the array
Space Complexity - O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE I

How many numbers are present in the array? 5 
Enter the numbers: 1 2 3 4 5
Which number do you want to search in the array? 6
The number 6 is not present in the array

SAMPLE II

How many numbers are present in the array? 3
Enter the numbers: 3 1 2
Which number do you want to search in the array? 2
The number 2 is present in the array.
 
"""
